# Exercise 02 -  OLAP Cubes - Grouping Sets

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [3]:
%%sql
SELECT sum(sales_amount) as revenue
FROM factsales

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [5]:
%%sql
SELECT c.country, sum(f.sales_amount) as revenue
FROM factsales as f
JOIN dimcustomer as c ON f.customer_key = c.customer_key
GROUP BY c.country

 * postgresql://student:***@127.0.0.1:5432/pagila_star
108 rows affected.


[('Cambodia', Decimal('191.47')),
 ('Turkey', Decimal('1662.12')),
 ('Germany', Decimal('831.04')),
 ('Chad', Decimal('135.68')),
 ('Madagascar', Decimal('93.78')),
 ('France', Decimal('374.04')),
 ('Iraq', Decimal('111.73')),
 ('Colombia', Decimal('709.41')),
 ('Japan', Decimal('3471.74')),
 ('American Samoa', Decimal('71.80')),
 ('Virgin Islands, U.S.', Decimal('122.68')),
 ('Vietnam', Decimal('746.28')),
 ('Kuwait', Decimal('111.74')),
 ('Ecuador', Decimal('390.13')),
 ('Nepal', Decimal('116.78')),
 ('Cameroon', Decimal('200.46')),
 ('Saudi Arabia', Decimal('523.79')),
 ('Bangladesh', Decimal('402.05')),
 ('Lithuania', Decimal('73.76')),
 ('Netherlands', Decimal('586.66')),
 ('Yugoslavia', Decimal('259.43')),
 ('Gambia', Decimal('129.70')),
 ('Nigeria', Decimal('1511.48')),
 ('Dominican Republic', Decimal('318.23')),
 ('Faroe Islands', Decimal('114.72')),
 ('Egypt', Decimal('694.39')),
 ('Zambia', Decimal('134.67')),
 ('United States', Decimal('4110.32')),
 ('Argentina', Decimal('1434.48')),
 ('Slovakia', Decimal('89.74')),
 ('Angola', Decimal('215.48')),
 ('Russian Federation', Decimal('3045.87')),
 ('Oman', Decimal('187.50')),
 ('Turkmenistan', Decimal('136.73')),
 ('Afghanistan', Decimal('67.82')),
 ('Czech Republic', Decimal('133.71')),
 ('Liechtenstein', Decimal('114.72')),
 ('Ethiopia', Decimal('91.77')),
 ('Azerbaijan', Decimal('245.43')),
 ('Brazil', Decimal('3200.52')),
 ('Israel', Decimal('422.01')),
 ('Senegal', Decimal('100.75')),
 ('Chile', Decimal('328.29')),
 ('New Zealand', Decimal('92.76')),
 ('Hungary', Decimal('111.71')),
 ('Mozambique', Decimal('344.20')),
 ('Mexico', Decimal('3307.04')),
 ('Finland', Decimal('101.74')),
 ('Taiwan', Decimal('1210.94')),
 ('Sri Lanka', Decimal('116.70')),
 ('Thailand', Decimal('419.04')),
 ('Estonia', Decimal('115.70')),
 ('Latvia', Decimal('262.40')),
 ('Venezuela', Decimal('683.30')),
 ('Peru', Decimal('468.88')),
 ('Holy See (Vatican City State)', Decimal('152.66')),
 ('Runion', Decimal('216.54')),
 ('Anguilla', Decimal('106.65')),
 ('Bahrain', Decimal('112.75')),
 ('Algeria', Decimal('383.10')),
 ('Malaysia', Decimal('369.15')),
 ('Iran', Decimal('950.75')),
 ('Spain', Decimal('606.58')),
 ('Ukraine', Decimal('730.42')),
 ('Tonga', Decimal('73.82')),
 ('South Korea', Decimal('574.65')),
 ('Belarus', Decimal('277.34')),
 ('French Guiana', Decimal('103.78')),
 ('Hong Kong', Decimal('142.70')),
 ('Malawi', Decimal('123.72')),
 ('Nauru', Decimal('148.69')),
 ('Romania', Decimal('235.38')),
 ('North Korea', Decimal('113.69')),
 ('Puerto Rico', Decimal('254.39')),
 ('Canada', Decimal('593.63')),
 ('China', Decimal('5802.73')),
 ('Austria', Decimal('307.22')),
 ('Moldova', Decimal('127.66')),
 ('South Africa', Decimal('1204.15')),
 ('Kenya', Decimal('253.46')),
 ('Poland', Decimal('877.97')),
 ('Kazakstan', Decimal('198.48')),
 ('United Kingdom', Decimal('924.80')),
 ('Brunei', Decimal('113.65')),
 ('Tunisia', Decimal('78.77')),
 ('Italy', Decimal('831.11')),
 ('Pakistan', Decimal('525.72')),
 ('Sweden', Decimal('144.66')),
 ('Armenia', Decimal('118.75')),
 ('Greece', Decimal('232.46')),
 ('India', Decimal('6630.27')),
 ('Yemen', Decimal('510.83')),
 ('Paraguay', Decimal('275.38')),
 ('Myanmar', Decimal('194.48')),
 ('Philippines', Decimal('2381.32')),
 ('Morocco', Decimal('307.29')),
 ('Greenland', Decimal('137.66')),
 ('Switzerland', Decimal('252.39')),
 ('Indonesia', Decimal('1510.33')),
 ('Tuvalu', Decimal('120.74')),
 ('Saint Vincent and the Grenadines', Decimal('96.75')),
 ('Bolivia', Decimal('183.53')),
 ('Congo, The Democratic Republic of the', Decimal('212.50')),
 ('Sudan', Decimal('227.46')),
 ('United Arab Emirates', Decimal('333.16')),
 ('Tanzania', Decimal('341.17')),
 ('French Polynesia', Decimal('235.46')),
 ('Bulgaria', Decimal('204.50'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [9]:
%%time
%%sql


SELECT d.month, sum(f.sales_amount) as revenue
FROM factsales as f
JOIN dimdate as d ON f.date_key = d.date_key
GROUP BY d.month

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 3.29 ms, sys: 509 µs, total: 3.8 ms
Wall time: 14 ms


[(1, Decimal('4824.43')),
 (3, Decimal('23886.56')),
 (4, Decimal('28559.46')),
 (2, Decimal('9631.88')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [10]:
%%sql
SELECT d.month, c.country, sum(f.sales_amount) as revenue
FROM factsales as f
JOIN dimdate as d ON f.date_key = d.date_key
JOIN dimcustomer as c ON f.customer_key = c.customer_key
GROUP BY d.month, c.country

 * postgresql://student:***@127.0.0.1:5432/pagila_star
484 rows affected.


[(1, 'Virgin Islands, U.S.', Decimal('0.99')),
 (4, 'Morocco', Decimal('100.74')),
 (2, 'Bahrain', Decimal('30.95')),
 (1, 'Zambia', Decimal('12.97')),
 (2, 'Oman', Decimal('32.92')),
 (1, 'Latvia', Decimal('12.97')),
 (1, 'United Kingdom', Decimal('73.84')),
 (2, 'Iraq', Decimal('12.97')),
 (2, 'Egypt', Decimal('121.73')),
 (5, 'United Arab Emirates', Decimal('4.99')),
 (3, 'Finland', Decimal('45.89')),
 (3, 'Chile', Decimal('149.70')),
 (4, 'Greece', Decimal('90.81')),
 (5, 'Mexico', Decimal('13.94')),
 (2, 'Mozambique', Decimal('14.95')),
 (3, 'Morocco', Decimal('126.72')),
 (1, 'China', Decimal('455.91')),
 (1, 'Faroe Islands', Decimal('6.99')),
 (1, 'Gambia', Decimal('14.97')),
 (1, 'Russian Federation', Decimal('225.40')),
 (1, 'Malaysia', Decimal('30.94')),
 (3, 'Greece', Decimal('82.80')),
 (4, 'Chile', Decimal('101.77')),
 (4, 'Finland', Decimal('32.90')),
 (1, 'Congo, The Democratic Republic of the', Decimal('43.92')),
 (3, 'Sri Lanka', Decimal('46.88')),
 (2, 'Armenia', Decimal('12.98')),
 (3, 'Vietnam', Decimal('278.35')),
 (2, 'Thailand', Decimal('63.84')),
 (3, 'Philippines', Decimal('905.98')),
 (3, 'India', Decimal('2547.23')),
 (3, 'Anguilla', Decimal('44.87')),
 (3, 'Holy See (Vatican City State)', Decimal('57.88')),
 (2, 'Dominican Republic', Decimal('33.91')),
 (4, 'Liechtenstein', Decimal('49.85')),
 (5, 'Germany', Decimal('14.95')),
 (3, 'South Korea', Decimal('177.64')),
 (1, 'Turkey', Decimal('120.73')),
 (2, 'Saint Vincent and the Grenadines', Decimal('15.96')),
 (2, 'Paraguay', Decimal('34.90')),
 (3, 'American Samoa', Decimal('7.96')),
 (1, 'Peru', Decimal('40.91')),
 (4, 'Bangladesh', Decimal('163.61')),
 (1, 'Nauru', Decimal('4.99')),
 (1, 'Italy', Decimal('51.89')),
 (2, 'Kuwait', Decimal('36.92')),
 (3, 'Iran', Decimal('350.22')),
 (4, 'Israel', Decimal('203.54')),
 (1, 'Tanzania', Decimal('18.95')),
 (4, 'United States', Decimal('1662.20')),
 (2, 'Nigeria', Decimal('177.60')),
 (3, 'Liechtenstein', Decimal('28.94')),
 (4, 'South Korea', Decimal('291.31')),
 (4, 'Anguilla', Decimal('37.86')),
 (4, 'Holy See (Vatican City State)', Decimal('46.87')),
 (2, 'Indonesia', Decimal('239.42')),
 (4, 'Philippines', Decimal('928.65')),
 (4, 'India', Decimal('2757.46')),
 (2, 'Venezuela', Decimal('140.68')),
 (4, 'Vietnam', Decimal('308.31')),
 (4, 'Sri Lanka', Decimal('45.88')),
 (3, 'United States', Decimal('1519.45')),
 (2, 'Sudan', Decimal('27.90')),
 (4, 'Iran', Decimal('410.05')),
 (3, 'Israel', Decimal('121.70')),
 (3, 'Bangladesh', Decimal('154.64')),
 (1, 'Azerbaijan', Decimal('20.97')),
 (5, 'Hungary', Decimal('0.99')),
 (4, 'American Samoa', Decimal('39.89')),
 (2, 'Chad', Decimal('9.98')),
 (5, 'United Kingdom', Decimal('11.96')),
 (2, 'Iran', Decimal('129.65')),
 (3, 'Kuwait', Decimal('53.87')),
 (4, 'Sudan', Decimal('105.76')),
 (3, 'Nigeria', Decimal('542.77')),
 (5, 'Latvia', Decimal('2.99')),
 (4, 'Chad', Decimal('48.89')),
 (3, 'Saint Vincent and the Grenadines', Decimal('17.94')),
 (5, 'Zambia', Decimal('0.99')),
 (3, 'Paraguay', Decimal('91.78')),
 (2, 'American Samoa', Decimal('23.95')),
 (4, 'Venezuela', Decimal('292.27')),
 (2, 'India', Decimal('832.92')),
 (2, 'Philippines', Decimal('359.17')),
 (4, 'Indonesia', Decimal('606.61')),
 (2, 'Holy See (Vatican City State)', Decimal('41.93')),
 (2, 'Anguilla', Decimal('16.95')),
 (1, 'Mexico', Decimal('263.37')),
 (3, 'Dominican Republic', Decimal('134.65')),
 (2, 'South Korea', Decimal('54.84')),
 (2, 'Sri Lanka', Decimal('10.97')),
 (3, 'Armenia', Decimal('31.93')),
 (2, 'Vietnam', Decimal('116.74')),
 (3, 'Thailand', Decimal('147.68')),
 (1, 'United Arab Emirates', Decimal('12.95')),
 (5, 'China', Decimal('40.87')),
 (4, 'Saint Vincent and the Grenadines', Decimal('46.88')),
 (4, 'Paraguay', Decimal('146.72')),
 (3, 'Chad', Decimal('63.85')),
 (2, 'United States', Decimal('590.48')),
 (4, 'Nigeria', Decimal('647.43')),
 (3, 'Sudan', Decimal('68.85')),
 (4, 'Kuwait', Decimal('14.97')),
 (2, 'Israel', Decimal('53.89')),
 (1, 'Pakistan', De

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [24]:
%%sql
SELECT d.month, c.country, sum(f.sales_amount) as revenue
FROM factsales as f
JOIN dimdate as d ON f.date_key = d.date_key
JOIN dimstore as c ON f.store_key = c.store_key
GROUP BY GROUPING SETS ((), d.month, c.country, (d.month, c.country))
ORDER BY revenue DESC
LIMIT 10

 * postgresql://student:***@127.0.0.1:5432/pagila_star
10 rows affected.


[(None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74')),
 (4, None, Decimal('28559.46')),
 (3, None, Decimal('23886.56')),
 (4, 'Canada', Decimal('14423.39')),
 (4, 'Australia', Decimal('14136.07')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (2, None, Decimal('9631.88'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>